### PR TITLE
docs: Extism Versioning System proposal

### PR DESCRIPTION
**Pull request by bhelx**
Rendered: https://github.com/extism/proposals/blob/fe24c6058e685a31ae846b8623f6dc4f0a919cfe/EIP-005-extism-versioning-system.md

### CODE REVIEW

File: EIP-005-extism-versioning-system.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#  Extism Versioning System
+
+## Purpose
+
+Extism has many separate release channels. There are three categories of software in the ecosystem:
+
+a. The Host SDKs
+b. The Plugin PDKs
+c. The runtime (the shared library)
```
Comment (nilslice) [inline EIP-005-extism-versioning-system.md:9]: I think these need a line between them if you want them to be on their own lines.

File: EIP-005-extism-versioning-system.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+2. Compatibility will be determined by a 2-word, codeword system. Each major version release of the runtime will get a new codeword. The first word will indicate Host compatibility and the second word will indicate Plugin compatibility. The codeword will be of the form `<html-color>-<sea-creature>`.
+
+The key to understanding this is to look at the ecosystem as having two coupling points, or two "APIs". The Host API and the Plugin API:
+
+![libextism coupling](content/005-coupling.png)
+
+Changes in the Host API will break the SDK libraries but not the PDKs and vice versa. We may encounter contradicitons to this but we should hopefylly be able to limit the changes to just the SDK. The Plugin API should rarely change as it breaks the entire world of built plugins. But the Host API may change more frequently. Thus we have decided to tie the version of the runtime to breaking changes in the Host API. Every new major bump in the runtime will represent a breaking change in the Host SDK. Rarely, a major bump in the runtime will indicate a breaking change in the Host OR the Plugin API.
```
Comment (nilslice) [inline EIP-005-extism-versioning-system.md:35]:
```suggestion
Changes in the Host API will break the SDK libraries but not the PDKs and vice versa. We may encounter contradicitons to this but we should hopefully be able to limit the changes to just the SDK. The Plugin API should rarely change as it breaks the entire world of built plugins. But the Host API may change more frequently. Thus we have decided to tie the version of the runtime to breaking changes in the Host API. Every new major bump in the runtime will represent a breaking change in the Host SDK. Rarely, a major bump in the runtime will indicate a breaking change in the Host OR the Plugin API. 
```

**Review by nilslice — CHANGES_REQUESTED** [review]
Just a minor typo w/ fix in a suggestion, and maybe a formatting change -- but looks great otherwise!

File: EIP-005-extism-versioning-system.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+The following is a hypothetical future state of the ecoystem:
+
+![Versioning Table](content/005-versioning.png)
+
+There are a few different methods we can explore to communicate compatibility. First I think we need documentation and to keep a table similar to the tables above. This should be generated from some parseable source of truth, like yaml.
+
+Second, I think we can encode into the SDK and PDK which versions they are compatible with. The way to do this would be to share a function b/w the Host API and the Plugin API. There is already a function which we could re-use in the host `const char *extism_version(void)`. We could also export this function into the Plugin API.
+
+The function would return a string like this `<semver-version>-<color>-<sea-creature>`. e.g.:
```
Comment (bhelx) [inline EIP-005-extism-versioning-system.md:47]: This makes the whole string valid semver. Maybe we don't want to do that an express these are 2 different things by instead having a format like `<semver-version>;<color>-<edition>`

Comment (nilslice) [reply EIP-005-extism-versioning-system.md]: its kind of nice that the whole string is semver, which has a side-effect of semver parsers being able to make sense of it?

**Review by nilslice — APPROVED** [review]
Super well thought out, nice work @bhelx. I think this is merge-worthy if you're ready. We can always make additional PRs to edit down the line.

**Pull request merged**